src/Test/QuickCheck/StateMachineModel.hs

```diff
@@ -349,132 +466,213 @@ linearise StateMachineModel {..} xs0 = anyP (step initialModel) . linearTree $ x
-  => History cmd ref -> Fork [Operation cmd ref]
-toForkOfOps h = Fork (fst $ mkOps n0 l) p' (fst $ mkOps n0 r)
+toForkOfOps :: forall cmd refs. History cmd refs -> Fork [Operation cmd refs]
+toForkOfOps h = Fork (mkOps l) p' (mkOps r)
   where
   (p, h') = partition (\e -> getProcessIdEvent e == 0) h
   (l, r)  = partition (\e -> getProcessIdEvent e == 1) h'
 
-  (p', n0) = mkOps 0 p
+  p'      = mkOps p
 
-  mkOps :: Int -> [HistoryEvent (Untyped cmd ref)] -> ([Operation cmd ref], Int)
-  mkOps n [] = ([], n)
-  mkOps n (InvocationEvent (Untyped cmd) _ : ResponseEvent resp pid : es)
-    | returnsRef (Untyped cmd) = let (ih, n') = mkOps (n + 1) es
-                       in  (Operation cmd (fromJust $ fromDynamic resp) pid (Just n) : ih, n')
-    | otherwise      = let (ih, n') = mkOps n es
-                       in  (Operation cmd (fromJust $ fromDynamic resp) pid Nothing  : ih, n')
-  mkOps _ _ = error "mkOps: Impossible."
+  mkOps :: [HistoryEvent (Untyped' cmd refs)] -> [Operation cmd refs]
+  mkOps [] = []
+  mkOps (InvocationEvent (Untyped' cmd miref) _ : ResponseEvent resp pid : es)
+    = Operation cmd (fromJust $ fromDynamic resp) pid : mkOps es
 
 ------------------------------------------------------------------------
 
-data HistoryKit cmd ref = HistoryKit
-  { getHistoryChannel   :: TChan (HistoryEvent (Untyped cmd ref))
+data HistoryKit cmd refs = HistoryKit
+  { getHistoryChannel   :: TChan (HistoryEvent (Untyped' cmd refs))
   , getProcessIdHistory :: Pid
   }
 
-mkHistoryKit :: Pid -> IO (HistoryKit cmd ref)
+mkHistoryKit :: Pid -> IO (HistoryKit cmd refs)
 mkHistoryKit pid = do
   chan <- newTChanIO
   return $ HistoryKit chan pid
 
 runMany
-  :: RefKit cmd
-  => Typeable ref
-  => HistoryKit cmd IntRef
-  -> (forall resp. cmd resp ref -> IO resp)
-  -> [Untyped cmd IntRef] -> StateT (Map IntRef ref) IO ()
-runMany kit step = flip foldM () $ \_ cmd'@(Untyped cmd) -> do
+  :: SDecide ix
+  => IxFunctor1 cmd
+  => HistoryKit cmd (ConstSym1 IntRef)
+  -> (forall resp. cmd resp refs -> IO (Response_ refs resp))
+  -> (forall resp refs'. cmd resp refs' -> SResponse ix resp)
+  -> [Untyped' cmd (ConstSym1 IntRef)]
+  -> StateT (IxMap ix IntRef refs) IO ()
+runMany kit step returns = flip foldM () $ \_ cmd'@(Untyped' cmd iref) -> do
   lift $ atomically $ writeTChan (getHistoryChannel kit) $
     InvocationEvent cmd' (getProcessIdHistory kit)
-  resp <- liftSem step (getProcessIdHistory kit) cmd
+  resp <- liftSem step returns cmd iref
+
   lift $ do
     threadDelay =<< randomRIO (0, 20)
     atomically $ writeTChan (getHistoryChannel kit) $
       ResponseEvent (toDyn resp) (getProcessIdHistory kit)
 
 parallelProperty
-  :: forall cmd ref model
-  .  (Ord (Untyped cmd IntRef), Show (Untyped cmd IntRef))
-  => Typeable ref
-  => Show (model IntRef)
-  => RefKit cmd
+  :: forall
+     (ix    :: *)
+     (cmd   :: Response ix -> (TyFun ix * -> *) -> *)
+     (refs  :: TyFun ix * -> *)
+     (model :: (TyFun ix * -> *) -> *)
```

Comment:
`stylish-haskell` doesn't seem to like `ix ~> *`, but it could be because of `TypeOperators` not being enabled; I'll try!